File: typescript-sdk/tests/test_determinism_optout.ts
/**
 * settle-capture (4b): Session.create({ noDeterminism }) forwards the
 * `no_determinism` flag to the session.create RPC. Default sessions omit it.
 *
 * Cross-run determinism v2: Session.create({ clockAnchor }) forwards
 * `clock_anchor` (the fixed epoch-ms that pins the injected browser clock).
 * Default sessions omit it (wall-clock epoch, unchanged wire shape).
 */
import { test, describe, before, after } from "node:test";
import assert from "node:assert/strict";
import { Session } from "../src/index.js";
import { MockDaemon } from "./helpers/mock_daemon.js";

describe("settle-capture: --no-determinism session opt-out", () => {
  let daemon: MockDaemon;
  let lastCreateParams: Record<string, unknown> = {};
  let counter = 0;

  before(async () => {
    daemon = new MockDaemon();
    daemon.registerHandler("session.create", (params) => {
      lastCreateParams = params;
      counter += 1;
      return {
        session_id: `01TEST${String(counter).padStart(20, "0")}`,
        status: "active",
        created_at_ms: 0,
      };
    });
    await daemon.start();
  });

  after(async () => {
    await daemon.stop();
  });

  test("default session create omits no_determinism (determinism ON)", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    await s.close();
    assert.equal(lastCreateParams["no_determinism"], undefined);
  });

  test("noDeterminism: true forwards no_determinism to the RPC", async () => {
    const s = await Session.create({
      socketPath: daemon.socketPath,
      token: daemon.token,
      noDeterminism: true,
    });
    await s.close();
    assert.equal(lastCreateParams["no_determinism"], true);
  });

  test("default session create omits clock_anchor (wall-clock epoch)", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    await s.close();
    assert.equal(lastCreateParams["clock_anchor"], undefined);
  });

  test("clockAnchor forwards clock_anchor alongside seed", async () => {
    const s = await Session.create({
      socketPath: daemon.socketPath,
      token: daemon.token,
      seed: 42,
      clockAnchor: 1_700_000_000_000,
    });
    await s.close();
    assert.equal(lastCreateParams["seed"], 42);
    assert.equal(lastCreateParams["clock_anchor"], 1_700_000_000_000);
  });
});

File: python-sdk/tests/test_determinism_optout.py
"""
settle-capture (4b): Session.create(no_determinism=...) forwards the
`no_determinism` flag to the session.create RPC. Default sessions omit it.
Covers sync + async.

Cross-run determinism v2: Session.create(clock_anchor=...) forwards
`clock_anchor` (the fixed epoch-ms that pins the injected browser clock).
Default sessions omit it (wall-clock epoch, unchanged wire shape).
"""

from __future__ import annotations

import asyncio

import loom


def _capture_create(daemon) -> dict:
    state = {"params": None, "n": 0}

    def _create(params: dict) -> dict:
        state["params"] = params
        state["n"] += 1
        return {
            "session_id": f"01TEST{state['n']:020}",
            "status": "active",
            "created_at_ms": 0,
        }

    daemon.register_handler("session.create", _create)
    return state


def test_default_create_omits_no_determinism(daemon):
    state = _capture_create(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token):
        pass
    assert "no_determinism" not in state["params"]


def test_no_determinism_true_is_forwarded(daemon):
    state = _capture_create(daemon)
    with loom.Session.create(
        socket_path=str(daemon.socket_path), token=daemon.token, no_determinism=True
    ):
        pass
    assert state["params"]["no_determinism"] is True


def test_async_no_determinism_true_is_forwarded(daemon):
    state = _capture_create(daemon)

    async def _run() -> None:
        s = await loom.AsyncSession.create(
            socket_path=str(daemon.socket_path), token=daemon.token, no_determinism=True
        )
        await s.close()

    asyncio.run(_run())
    assert state["params"]["no_determinism"] is True


def test_default_create_omits_clock_anchor(daemon):
    state = _capture_create(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token):
        pass
    assert "clock_anchor" not in state["params"]


def test_clock_anchor_is_forwarded_with_seed(daemon):
    state = _capture_create(daemon)
    with loom.Session.create(
        socket_path=str(daemon.socket_path),
        token=daemon.token,
        seed=42,
        clock_anchor=1_700_000_000_000,
    ):
        pass
    assert state["params"]["seed"] == 42
    assert state["params"]["clock_anchor"] == 1_700_000_000_000


def test_async_clock_anchor_is_forwarded(daemon):
    state = _capture_create(daemon)

    async def _run() -> None:
        s = await loom.AsyncSession.create(
            socket_path=str(daemon.socket_path),
            token=daemon.token,
            clock_anchor=1_700_000_000_000,
        )
        await s.close()

    asyncio.run(_run())
    assert state["params"]["clock_anchor"] == 1_700_000_000_000
